module/VuFind/src/VuFind/Db/Service/OaiResumptionService.php

```diff
@@ -220,4 +229,45 @@ protected function encodeParams(array $params): string
         $processedParams = http_build_query($params, '', '&', PHP_QUERY_RFC3986);
         return $processedParams;
     }
+
+    /**
+     * Delete expired records. Allows setting a limit so that rows can be deleted in small batches.
+     *
+     * @param DateTime $dateLimit Date threshold of an "expired" record.
+     * @param ?int     $limit     Maximum number of rows to delete or null for no limit.
+     *
+     * @return int Number of rows deleted
+     */
+    public function deleteExpired(DateTime $dateLimit, ?int $limit = null): int
+    {
+        $subQueryBuilder = $this->entityManager->createQueryBuilder();
+        $subQueryBuilder->select('o.id')
+            ->from(OaiResumptionEntityInterface::class, 'o')
+            ->where('o.expires < :dateLimit')
+            ->setParameter('dateLimit', $dateLimit);
+
+        if ($limit) {
+            $subQueryBuilder->setMaxResults($limit);
+        }
+        $queryBuilder = $this->entityManager->createQueryBuilder();
+        $queryBuilder->delete(OaiResumptionEntityInterface::class, 'o')
+            ->where('o.id IN (:tokens)')
+            ->setParameter('tokens', $subQueryBuilder->getQuery()->getResult());
+
+        return $queryBuilder->getQuery()->execute();
+    }
+
+    /**
+     * Get current time or specified time as DateTime.
+     *
+     * This wrapper method makes testing easier.
+     *
+     * @param string $datetime Time to return
+     *
+     * @return DateTime
+     */
+    protected function getDateTime(string $datetime = 'now'): DateTime
+    {
+        return new Datetime($datetime);
```

Comment:
If we do keep this, we should fix inconsistent capitalization here:


```suggestion
        return new DateTime($datetime);
```